Python/ceval.c

```diff
@@ -2476,14 +2476,42 @@ _PyEval_GetBuiltinId(_Py_Identifier *name)
 PyObject *
 PyEval_GetLocals(void)
 {
+    // We need to return a borrowed reference here, so some tricks are needed
     PyThreadState *tstate = _PyThreadState_GET();
      _PyInterpreterFrame *current_frame = _PyThreadState_GetFrame(tstate);
     if (current_frame == NULL) {
         _PyErr_SetString(tstate, PyExc_SystemError, "frame does not exist");
         return NULL;
     }
 
-    PyObject *locals = _PyEval_GetFrameLocals();
+    // Be aware that this returns a new reference
+    PyObject *locals = _PyFrame_GetLocals(current_frame);
+
+    if (locals == NULL) {
+        return NULL;
+    }
+
+    if (PyFrameLocalsProxy_Check(locals)) {
+        PyFrameObject *f = _PyFrame_GetFrameObject(current_frame);
+        PyObject *ret = PyDict_New();
+        if (ret == NULL) {
+            Py_DECREF(locals);
+            return NULL;
+        }
+        if (PyDict_Update(ret, locals) < 0) {
+            Py_DECREF(ret);
+            Py_DECREF(locals);
+            return NULL;
+        }
+        Py_DECREF(locals);
+        Py_XSETREF(f->f_locals_cache, ret);
+        return ret;
```

Comment:
```suggestion
        PyObject *ret = f->f_locals_cache;
        if (ret == NULL)
            ret = PyDict_New();
            if (ret == NULL) {
                Py_DECREF(locals);
                return NULL;
            }
            Py_XSETREF(f->f_locals_cache, ret);
        }
        if (PyDict_Update(ret, locals) < 0) {
            Py_DECREF(ret);
            Py_DECREF(locals);
            return NULL;
        }
        Py_DECREF(locals);
        return ret;
```

Reply:
Suggestion updated to store returned snapshots in a list, so the function can switch to the semantics of `locals()` while still returning a borrowed reference that remains valid for the life of the frame object.

Reply:
Also adding back the alternate suggestion to keep the Python 3.12 `locals` semantics for this one API (since existing users are familiar with it anyway, and it works better with the borrowed reference than the Python 3.13 semantics do):

```suggestion
        PyObject *ret = f->f_locals_cache;
        if (ret == NULL) {
            PyObject *ret = PyDict_New();
            if (ret == NULL) {
                Py_DECREF(locals);
                return NULL;
            }
            f->f_locals_cache = cache;
        }
        if (PyDict_Update(ret, locals) < 0) {
            // At this point, if the cache dict is broken, it will stay broken, as
            // trying to clean it up or replace it will just cause other problems
            Py_DECREF(locals);
            return NULL;
        }
        Py_DECREF(locals);
        return ret;
```